### DISCUSSION TITLE
Job/Trigger Listener and DB/Data Context Dependency Injection

### DISCUSSION THREAD

**Discussion by awdorrin** [quartznet/quartznet#1725] [Q&A]
I am trying to inject an instance of my database context into a custom listener class, so I can make custom log updates to a database handle while listening to the results of Jobs and Trigger events.

I created my Listener class like this:
```
    public class QuartzListener : IJobListener, ITriggerListener
    {
        public string Name => "AllJobMonitor";
        private readonly DataContext _dbContext;
        private readonly EmailService _emailService;
        public readonly IConfiguration _config;

        public QuartzListener( DataContext dbContext, EmailService emailService, IConfiguration config)
        {
            _dbContext = dbContext;
            _emailService = emailService;
            _config = config;
        }
```
However, when I try to add the Job land Trigger Listeners in Startup.cs ConfigureServices I get an exception.
```
    q.AddJobListener<QuartzListener>(GroupMatcher<JobKey>.AnyGroup());
    q.AddTriggerListener<QuartzListener>(GroupMatcher<TriggerKey>.AnyGroup());
```
```
2>System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
2> ---> System.AggregateException: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Quartz.IJobListener Lifetime: Singleton ImplementationType: MHSCo.QuartzJobs.QuartzListener': Cannot consume scoped service 'Data.Models.DataContext' from singleton 'Quartz.IJobListener'.)
2> ---> System.InvalidOperationException: Error while validating the service descriptor 'ServiceType: Quartz.IJobListener Lifetime: Singleton ImplementationType: MHSCo.QuartzJobs.QuartzListener': Cannot consume scoped service 'Data.Models.DataContext' from singleton 'Quartz.IJobListener'.
2> ---> System.InvalidOperationException: Cannot consume scoped service 'Data.Models.DataContext' from singleton 'Quartz.IJobListener'.
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitScopeCache(ServiceCallSite scopedCallSite, CallSiteValidatorState state)
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.VisitConstructor(ConstructorCallSite constructorCallSite, CallSiteValidatorState state)
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
2>   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateCallSite(ServiceCallSite callSite)
2>   at Microsoft.Extensions.DependencyInjection.ServiceProvider.ValidateService(ServiceDescriptor descriptor)
```

I've been searching the docs and google trying to see if anyone has done something like this before, but I'm not having any luck.
Hoping someone may be able to point out what I'm doing wrong.
Thanks!

**Comment by lahma** [comment]
Listeners in Quartz are scheduler global, so they cannot consume more short-lived services. I think you need to inject `IServiceProvider` and get the db context from there for the duration of listener invocation (method scope).

**Reply by awdorrin** [reply]
Hmm, ok, after refreshing my memory on the difference between singleton, scoped and transient, I think I understand what you are saying.
I guess I never realized that services.AddDbContext() is creating a scoped service for the database context.
I think this gets me pointed in the right direction, thanks!

**Reply by xantari** [reply]
This seems to imply we can't inject DbContext's this way.  I tried it using UseMicrosoftDependencyInjectionJobFactory(), since UseMicrosoftDependencyInjectionScopedJobFactory(0 is now deprecated, and while I can get my DbContext injected via constructor injection, it doesn't work well when multiple triggers are firing the same IJob because then you get the following error:

System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.

**Reply by xantari** [reply]
Changed all my code to use IServiceProvider to fetch services instead, and still get the same threading issue with DbContext interactions... Can't for the life of me figure out how to get this to work with IJob when multiple triggers are firing at the same time for a single IJob.
